syntax = "proto3";

service RaftService {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntryRequest) returns (AppendEntriesResponse);
    rpc GetLeader(EmptyRequest) returns (GetLeaderResponse);
    rpc Suspend(SuspendRequest) returns (SuspendResponse);
    rpc SetVal(SetValRequest) returns (SetValResponse);
    rpc GetVal(GetValRequest) returns (GetValPesponse);
}

message RequestVoteRequest {
    int32 term = 1;
    int32 candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteResponse {
    int32 term = 1;
    bool result = 2;
}

message AppendEntryRequest {
    int32 term = 1;
    int32 leaaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated Entry entries = 5;
    int32 leaderCommit = 6;
}

message Entry {
    int32 id = 1;
    int32 termNumber = 2;
    string message = 3;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

message EmptyRequest{
}

message SuspendResponse {
    string message = 1;
}

message GetLeaderResponse{
    int32 leaderId = 1;
    string address = 2;
}

message SuspendRequest {
    int32 period = 1;
}

message GetValRequest {
    string key = 1;
}

message GetValPesponse {
    bool success = 1;
    string value = 2;
}

message SetValRequest {
    string key = 1;
    string value = 2;
}

message SetValResponse {
    bool success = 1;
}

